crates/ghost_actor/src/ghost_mutex.rs

```diff
@@ -0,0 +1,197 @@
+extern crate lock_api;
+extern crate parking_lot;
+
+use std::{
+    borrow::{Borrow, BorrowMut},
+    convert::{AsMut, AsRef},
+    ops::{Deref, DerefMut},
+};
+
+#[must_use]
+pub struct GhostMutexGuard<'lt, T: ?Sized + 'lt> {
+    guard: Option<parking_lot::MutexGuard<'lt, T>>,
+}
+
+impl<'lt, T: ?Sized + 'lt> Drop for GhostMutexGuard<'lt, T> {
+    fn drop(&mut self) {
+        match std::mem::replace(&mut self.guard, None) {
+            Some(guard) => {
+                parking_lot::MutexGuard::unlock_fair(guard);
+            }
+            None => (),
+        }
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> Deref for GhostMutexGuard<'lt, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &*self.guard.as_ref().expect("exists")
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> DerefMut for GhostMutexGuard<'lt, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut *self.guard.as_mut().expect("exists")
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> AsRef<T> for GhostMutexGuard<'lt, T> {
+    fn as_ref(&self) -> &T {
+        self.deref()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> AsMut<T> for GhostMutexGuard<'lt, T> {
+    fn as_mut(&mut self) -> &mut T {
+        self.deref_mut()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> Borrow<T> for GhostMutexGuard<'lt, T> {
+    fn borrow(&self) -> &T {
+        self.deref()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> BorrowMut<T> for GhostMutexGuard<'lt, T> {
+    fn borrow_mut(&mut self) -> &mut T {
+        self.deref_mut()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt + std::fmt::Debug> std::fmt::Debug for GhostMutexGuard<'lt, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if f.alternate() {
+            write!(f, "{:#?}", self.deref())
+        } else {
+            write!(f, "{:?}", self.deref())
+        }
+    }
+}
+
+pub struct GhostMutex<T: ?Sized> {
+    mutex: parking_lot::Mutex<T>,
+}
+
+impl<T> GhostMutex<T> {
+    pub fn new(t: T) -> Self {
+        Self {
+            mutex: parking_lot::Mutex::new(t),
+        }
+    }
+
+    pub fn into_inner(self) -> T {
+        self.mutex.into_inner()
+    }
+}
+
+impl<T: ?Sized> GhostMutex<T> {
+    pub fn lock(&self) -> GhostMutexGuard<'_, T> {
+        match self
+            .mutex
+            .try_lock_for(std::time::Duration::from_millis(10))
```

Comment:
That's a small timeout no ? Could you tell me how you found this to be optimal in your opinion ?

Reply:
For the moment, we are running in single threaded mode, so the only time a lock should be a problem is if we lock it twice from the same thread... in which case we want it to panic! so we can fix the circular dependency. But 1 second would be fine from a human time perspective and that might be more appropriate if we ever went multi-threaded too... will update.